### DISCUSSION TITLE
Question About Using Multiplexed Redis in Actix Web

### DISCUSSION THREAD

**Discussion by tyrsday29** [actix/actix-web#3461] [Q&A] [answered]
Hi!

Hoping someone has some experience in this area an knows the correct method of multiplexing a Redis connection in Actix Web. I'm currently using `ThinData<MultiplexedConnection>` passed to handlers which I clone to use, but am realizing that sharing that piece of data across threads even when cloning to use it may not be good practice, and nothing about `ThinData<T>` that I can see leads me to believe  it's cloning it. Probably the opposite do to the initial sentence of the docs about using internal arcs, sharing data some other way, or being static and cheap to clone (not exactly sure which category a multiplexed connection would fall in either).

 Anyway just hoping someone could guide me on where or not I should be making a struct to hold the MutltiplexedConnection in an Arc, and  then cloning the data out of the struct (instead of what I'm doing putting the connection in there directly), or at least whether my thinking is correct about `ThinData<T>`

**Comment by robjtede** [comment, marked as answer]
`ThinData` is for database pools and other types which share state internally, via Arc or whatever other method.

Key point from the `ThinData` docs:

> Unlike `Data`, this wrapper clones `T` during extraction. Therefore, it is the user’s responsibility to ensure that clones of `T` do actually share the same state, otherwise state may be unexpectedly different across multiple requests.

I'd probably still recommend putting the `redis::Client` in app data and not the connection itself, but using `client.get_multiplexed_async_connection().await` when you need it.

**Reply by tyrsday29** [reply]
thank you! I appreciate the response, I probably would have continued to do that wrong.
